celery/concurrency/asynpool.py

```diff
@@ -1085,6 +1086,11 @@ def on_job_process_lost(self, job, pid, exitcode):
         was assigned to exited by mysterious means (error exitcodes and
         signals).
         """
+        # if exit with EX_OK, won't mark it as worker lost,
```

Comment:
If I am not mistaken, at this point it is not clear whether the process exited **after the job was completed**. Perhaps you also need to check if the `job.ready` [method](https://github.com/celery/billiard/blob/13a52dc1c3dd9ba2f01c255064b33ceb71a9cb85/billiard/pool.py#L1136) returns `True`.

Reply:
Yes, it's really important to make sure the job is completes before the process exits (which is the warm shutdown case) !

I will improve it, thanks ！

Reply:
I tested again and found that job.ready() is False, but exitcode is EX_OK, is it a expected state ?

I think the job process exit with EX_OK indicates that the job it executes is also OK, any suggestions ?

> If I am not mistaken, at this point it is not clear whether the process exited **after the job was completed**. Perhaps you also need to check if the `job.ready` [method](https://github.com/celery/billiard/blob/13a52dc1c3dd9ba2f01c255064b33ceb71a9cb85/billiard/pool.py#L1136) returns `True`.

Reply:
@tinylambda I am quite certain that it is not necessary. A process can shutdown gracefully without necessarily completing the task.

Reply:
Please rephrase the comment.
"If the process exists with the EX_OK status code, don't mark the worker as lost"
I think it's clearer that way.
Also,for future reference please mention the relevant issue number.